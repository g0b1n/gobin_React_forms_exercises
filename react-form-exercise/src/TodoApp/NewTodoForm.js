import React, {useState} from "react";

const NewTodoForm = ({addTodo}) => {    
    const InitialState = {
        todo: ""
    }

    const [formData, setFormData] = useState(InitialState);
    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        addTodo({...formData})
        setFormData(InitialState)
    }

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="todo">Todo</label>
            <input 
                id="todo"
                type="text" name="todo"
                placeholder="Add new todo"
                value={formData.todo}
                onChange={handleChange}
            />

            <button>Add New Todo</button>
        </form>
    )
}

export default NewTodoForm;