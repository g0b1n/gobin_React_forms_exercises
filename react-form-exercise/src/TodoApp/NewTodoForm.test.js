import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewTodoForm from "./NewTodoForm";

// smoke test
test('it renders without crashing', () => {
    render(<NewTodoForm />)
})

// snapshot test
test('it matches the snapshot', () => {
    const {asFragment} = render(<NewTodoForm />);
    expect(asFragment()).toMatchSnapshot()
})

it("should add new todo", () => {
    const addTodo = jest.fn();

    const { getByText, getByLabelText } = render(<NewTodoForm addTodo={addTodo} />);
    const todoInput = getByLabelText("Todo");

    fireEvent.change(todoInput, {target: {value: "Clean my room"}})

    fireEvent.click(getByText('Add New Todo'));

    expect(addTodo).toHaveBeenCalledWith({
        todo: 'Clean my room'
    });
})