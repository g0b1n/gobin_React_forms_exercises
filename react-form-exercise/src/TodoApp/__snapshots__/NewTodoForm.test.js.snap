// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="todo"
    >
      Todo
    </label>
    <input
      id="todo"
      name="todo"
      placeholder="Add new todo"
      type="text"
      value=""
    />
    <button>
      Add New Todo
    </button>
  </form>
</DocumentFragment>
`;
